File: frontend/src/services/productService.ts
import api from './api';
import { Product, ProductRequest } from '../types';

export const productService = {
    // Get all products
    getAll: async (): Promise<Product[]> => {
        const response = await api.get<Product[]>('/api/products');
        return response.data;
    },

    // Get product by ID
    getById: async (id: string): Promise<Product> => {
        const response = await api.get<Product>(`/api/products/${id}`);
        return response.data;
    },

    // Create new product (ADMIN only)
    create: async (product: ProductRequest): Promise<Product> => {
        const response = await api.post<Product>('/api/products', product);
        return response.data;
    },

    // Update product (ADMIN only)
    update: async (id: string, product: ProductRequest): Promise<Product> => {
        const response = await api.put<Product>(`/api/products/${id}`, product);
        return response.data;
    },

    // Delete product (ADMIN only)
    delete: async (id: string): Promise<void> => {
        await api.delete(`/api/products/${id}`);
    },
};

File: server/product-service/src/main/java/com/example/productservice/services/ProductService.java
package com.example.productservice.services;

import com.example.productservice.entities.Product;
import com.example.productservice.repositories.ProductRepository;
import jakarta.transaction.Transactional;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@AllArgsConstructor
public class ProductService {
    private final ProductRepository productRepository;

    // Add a product
    public Product create(Product product) {
        return productRepository.save(product);
    }

    // Edit a product
    @Transactional
    public Product update(String id, Product product) {
        return productRepository.findById(id)
                .map(existingProduct -> {
                    existingProduct.setName(product.getName());
                    existingProduct.setDescription(product.getDescription());
                    existingProduct.setPrice(product.getPrice());
                    existingProduct.setQuantity(product.getQuantity());
                    return productRepository.save(existingProduct);
                })
                .orElseThrow(() -> new RuntimeException("Product with id " + id + " does not exist"));
    }

    // List all products
    public List<Product> getAll() {
        return productRepository.findAll();
    }

    // Details of product using id
    public Product getById(String id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));
    }

    // Delete product using id
    public void delete(String id) {
        if (!productRepository.existsById(id)) {
            throw new RuntimeException("Cannot delete: Product with id " + id + " does not exist");
        }
        productRepository.deleteById(id);
    }
}
